hypervisor/src/kvm/mod.rs

```diff
@@ -696,9 +696,35 @@ impl cpu::Vcpu for KvmVcpu {
         match self.fd.run() {
             Ok(run) => match run {
                 #[cfg(target_arch = "x86_64")]
-                VcpuExit::IoIn(addr, data) => Ok(cpu::VmExit::IoIn(addr, data)),
+                VcpuExit::IoIn(addr, data) => {
+                    if let Some(ref vmmops) = self.vmmops {
+                        let ret = vmmops.lock().unwrap().pio_read(addr.into(), data);
+                        if ret.is_ok() {
+                            return Ok(cpu::VmExit::Ignore);
+                        }
+                    }
+                    Err(cpu::HypervisorCpuError::RunVcpu(anyhow!(
```

Comment:
In the original handler, a warning is thrown if an IO address cannot be read/written to. While using vmmops I modified it to return an error. With this change unit tests seem to pass just fine. 

Can someone confirm if (or not) warning on read/write is the correct approach here?

Reply:
It depends on whether you want the VCPU to get terminated or not in that situation.

Since your goal doesn't not include changing the failure mode, I would suggest you retain the original behaviour. That is, only warn but don't return an error.